lib/hsa/kernels/frb_beamform_amd.cl

```diff
@@ -0,0 +1,259 @@
+    __local float2 transfer[512];
+    #pragma unroll
+    for (int ew=0; ew<4; ew++) { //cylinder
+        //de-mux cyl into local share
+        uint irev = L*8;;
+        __asm__ ("V_BFREV_B32 %0, %1" : "=v"(irev) : "v"(irev)); //32b bit-reverse
+        irev = irev>>23; //9b reverse
+        transfer[irev    ] = res[ew][0]; //hardcoded offsets to speed up bit flip
+        transfer[irev+256] = res[ew][1];
+        transfer[irev+128] = res[ew][2];
+        transfer[irev+384] = res[ew][3];
+        transfer[irev+ 64] = res[ew][4];
+        transfer[irev+320] = res[ew][5];
+        transfer[irev+192] = res[ew][6];
+        transfer[irev+448] = res[ew][7];
+
+        barrier(CLK_LOCAL_MEM_FENCE);
+        //fetch the right ns beams for cyl
+        #pragma unroll
+        for (int ns=0; ns<4; ns++){
+            uint addr = map[L*4+ns];
+            res[ew][ns] = transfer[addr];
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    //form beams
+    float2 beam[4][4];
+    #pragma unroll
+    for (int bidx=0; bidx<4; bidx++){ //4 beams EW
+        #pragma unroll
+        for (int ns=0; ns<4; ns++){
+            beam[ns][bidx].RE = (res[0][ns].RE*lco[bidx*4+0].RE + res[0][ns].IM*lco[bidx*4+0].IM +
+                                 res[1][ns].RE*lco[bidx*4+1].RE + res[1][ns].IM*lco[bidx*4+1].IM +
+                                 res[2][ns].RE*lco[bidx*4+2].RE + res[2][ns].IM*lco[bidx*4+2].IM +
+                                 res[3][ns].RE*lco[bidx*4+3].RE + res[3][ns].IM*lco[bidx*4+3].IM)/4.f;
+            beam[ns][bidx].IM = (res[0][ns].RE*lco[bidx*4+0].IM - res[0][ns].IM*lco[bidx*4+0].RE +
+                                 res[1][ns].RE*lco[bidx*4+1].IM - res[1][ns].IM*lco[bidx*4+1].RE +
+                                 res[2][ns].RE*lco[bidx*4+2].IM - res[2][ns].IM*lco[bidx*4+2].RE +
+                                 res[3][ns].RE*lco[bidx*4+3].IM - res[3][ns].IM*lco[bidx*4+3].RE)/4.f;
+            uint addr = T * 2048 + //time
+                        P * 1024 + //pol
+                        bidx * 256 +             //EW
+                        (255-(L*4+ns));          //position
+//            vstore_half2_rtz(beam[ns][bidx],0,(half*)&outputData[addr]);    //Slooower (substantially)
+            outputData[addr] = (half2){beam[ns][bidx].RE,beam[ns][bidx].IM};  //Slower (by a hair)
+//            float out; //FAST [only on newer rocms...?]
+//            __asm__ ("V_CVT_PKRTZ_F16_F32 %0, %1, %2" : "=v"(out): "v"(beam[ns][bidx].RE), "v"(beam[ns][bidx].IM));
+//            outputData[addr] = out;
+        }
+    }
+}
+
+
+
+/// tooling around
+/*            addr = (T/6144) * 1024 * (6144+256) +     // 0..8 T    ZZZ
+                   (L*4 + ns) * 6144 +                // 0..256 NS WG
+                   bidx * 256 * 6144 +                // 0..4 EW   WG
+                   P * 1024 * (6144+256) +                  // 0..1 Pol  YYY
+                   (T%6144);                          // 0..6144 t ZZZ
+            */
```

Comment:
Remove this and other commented tests from production code, unless we think it'll be useful later?  If so, let's add more comments.

Reply:
Definitely removable, as the earlier testing stuff.